files/chef-marketplace-cookbooks/chef-marketplace/libraries/helpers.rb

```diff
@@ -386,6 +391,16 @@ def biscotti_token_hmac
                               node["chef-marketplace"]["biscotti"]["token"],
                               node["chef-marketplace"]["biscotti"]["uuid"])
     end
+
+    def download_url(product)
```

Comment:
In principal I like this idea assuming we're forced to put the packages in their cloud because of firewall issues. I think that we need to formarlize and unify our approach in this and the image generation.
1) Both the upgrade and marketplace_image install should use the same mechanism for determining available package versions.
2) Perhaps we should have a `latest` version that points to the latest version for each software package.
3) Docs and/or code to upload packages and modify that config file should be present.

Alternatively, have we considered mirroring the repos in Alibaba?

Reply:
It is a speed issue when downloading packages hence the reason for using their storage service.
For example when I did this manually it would take almost all day to download the marketplace image!

I agree with the same method for all scenarios - just trying to get it to work in the first place ;-)

The _easiest_ approach would be to mirror into the Alibaba repos. This would keep everything as is for other cloud platforms - just need to take out the specifics for Alibaba that I have put in.

Reply:
I do not mean the main Ubuntu repos, it would be good if we could get a `packages` repo in China.

Reply:
@schisamo, fyi about downloads from `packages.chef.io` in China.

Reply:
@yzl There should be no technical restriction on mirroring packages.chef.io data in China, it is not on our embargoed country list:
https://github.com/chef/es-package-router/blob/05f8910a7a77904e02b4ad40e7b493c717c3ce4f/.delivery/build-cookbook/recipes/_fastly.rb#L3-L9

Reply:
@schisamo @ryancragun  @yzl Interesting.

I am not sure what the issue is then, but there is very definite throttling going on when trying to get a package from `packages.chef.io`. It is so slow that it more often than not times out.

I am hoping to get some contact details for Alibaba as there is another issue I have with the approval of the image which has been rejected. (The error states that we running Windows Update will fix it!!!)

Reply:
I read about this a bit, and the possibilities that come to my mind are:

1. The package isn't cached at whatever fastly POP is closest to Alibaba so they're actually downloading from the origin in the US,
2. The package is cached locally, but we need some support from fastly, e.g. https://community.fastly.com/t/fastly-cdn-content-blocked-by-the-great-firewall-of-china/447
3. The package is cached locally and fastly is set up for our particular use case, and something else is going on.